docs/hazmat/primitives/asymmetric/dsa.rst

```diff
@@ -74,8 +74,55 @@ provider.
     ...     key_size=1024,
     ...     backend=default_backend()
     ... )
-    >>> signer = private_key.signer(hashes.SHA256())
     >>> data = b"this is some data I'd like to sign"
+
+Signing using a :class:`cryptography.hazmat.primitives.hashes.HashContext` (preferred method):
+
+.. doctest::
+
+    >>> digester = hashes.Hash(hashes.SHA256(), backend=default_backend())
+    >>> digester.update(data)
+    >>> signature = private_key.sign(digester)
+
+One-line signing:
+
+.. doctest::
+
+    >>> signature = private_key.sign(data, hashes.SHA256())
+
+Signing a precomputed digest:
+
+.. warning::
+
+    Only use this for digests, do NOT sign raw data. The data will get truncated
+    to the size of DSA's "q" parameter, which is just a few bytes.
+
+.. doctest::
+
+    >>> digester = hashes.Hash(hashes.SHA256(), backend=default_backend())
+    >>> digester.update(data)
+    >>> digest = digester.finalize()
+    >>> signature = private_key.sign(digest, already_hashed=True)
+
+Specifying the per-message key k:
```

Comment:
I would say setting k is out of scope for this API concern (and separately,
we should never support it)

On Wed, Jan 6, 2016 at 12:49 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/hazmat/primitives/asymmetric/dsa.rst
> <https://github.com/pyca/cryptography/pull/2642#discussion_r48986337>:
>
> > +
> > +Signing a precomputed digest:
> > +
> > +.. warning::
> > +
> > +    Only use this for digests, do NOT sign raw data. The data will get truncated
> > +    to the size of DSA's "q" parameter, which is just a few bytes.
> > +
> > +.. doctest::
> > +
> > +    >>> digester = hashes.Hash(hashes.SHA256(), backend=default_backend())
> > +    >>> digester.update(data)
> > +    >>> digest = digester.finalize()
> > +    >>> signature = private_key.sign(digest, already_hashed=True)
> > +
> > +Specifying the per-message key k:
>
> Under what scenario do you want the ability to set k manually? I realize
> there's a warning below, but this API is incredibly dangerous.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2642/files#r48986337>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084